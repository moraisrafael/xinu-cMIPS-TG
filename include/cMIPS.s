
	# see vhdl/packageMemory.vhd for addresses
        .set x_INST_BASE_ADDR,0x00000000
        .set x_INST_MEM_SZ,0x00020000

        .set x_DATA_BASE_ADDR,0x00080000
        .set x_DATA_MEM_SZ,0x00080000
	
        .set x_IO_BASE_ADDR,0x3c000000
        .set x_IO_MEM_SZ,0x00002000
	.set x_IO_ADDR_RANGE,0x00000020

	.set x_SDRAM_BASE_ADDR,0x04000000
	.set x_SDRAM_MEM_SZ,0x02000000
	
	.set HW_counter_addr,(x_IO_BASE_ADDR +  5 * x_IO_ADDR_RANGE)
	.set HW_FPU_addr,    (x_IO_BASE_ADDR +  6 * x_IO_ADDR_RANGE)
	.set HW_uart_addr,   (x_IO_BASE_ADDR +  7 * x_IO_ADDR_RANGE)
	.set HW_dsp7seg_addr,(x_IO_BASE_ADDR +  9 * x_IO_ADDR_RANGE)
	.set HW_keybd_addr,  (x_IO_BASE_ADDR + 10 * x_IO_ADDR_RANGE)
	.set HW_lcd_addr,    (x_IO_BASE_ADDR + 11 * x_IO_ADDR_RANGE)
	.set HW_sdc_addr,    (x_IO_BASE_ADDR + 12 * x_IO_ADDR_RANGE)

	# see vhdl/packageMemory.vhd for addresses
	.set x_EXCEPTION_0000,0x00000130
	.set x_EXCEPTION_0100,0x00000200
	.set x_EXCEPTION_0180,0x00000280
	.set x_EXCEPTION_0200,0x00000400
	.set x_EXCEPTION_BFC0,0x00000680
	.set x_ENTRY_POINT,   0x00000700


	.set c0_index,   $0
	.set c0_random,  $1
	.set c0_entrylo0,$2
	.set c0_entrylo1,$3
	.set c0_context ,$4
	.set c0_pagemask,$5
	.set c0_wired,   $6
	.set c0_badvaddr,$8
	.set c0_count   ,$9
	.set c0_entryhi ,$10
	.set c0_compare ,$11
	.set c0_status  ,$12
	.set c0_cause   ,$13
	.set c0_epc,     $14
	.set c0_config,  $16
	.set c0_config_f0,0
	.set c0_config_f1,1
	.set c0_lladdr,  $17
	.set c0_errorpc, $30

	
	# reset: COP0 present, at exception level, all else disabled
	.set c0_status_reset,0x10000002
	
	# normal state: COP0 present, user mode, all IRQs enabled
	.set c0_status_normal,0x1000ff11

	# reset: COUNTER stopped, use special interrVector, no interrupts
	.set c0_cause_reset, 0x0880007c


	# not used in XINU, RH
#	.equ led_RED, 0x4000
#	.equ led_GRE, 0x2000
#	.equ led_BLU, 0x1000
#	.equ led_OFF, 0x0FFF

