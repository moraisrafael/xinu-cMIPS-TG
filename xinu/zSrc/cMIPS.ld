OUTPUT_ARCH(mips)
ENTRY(_start)

MEMORY
{
        rom (rx)  : ORIGIN = 0x, /* x_INST_BASE_ADDR */
                    LENGTH = 0x, /* x_INST_MEM_SZ */
        ram (!rx) : ORIGIN = 0x, /* x_DATA_BASE_ADDR */
                    LENGTH = 0x, /* x_DATA_MEM_SZ */
        io (!rx)  : ORIGIN = 0x0f000000, /* not used, do not remove       */
                    LENGTH = 0x00020000  /* edMemory.sh needs thess lines */
}

SECTIONS
{
       .text :
       {
          *(.text .text.*)
          _etext = . ;         /* end of text constant (from Xinu) */
       }                                   > rom

       .data : 
       {
          _bdata = . ;         /* begin of data constant */
          *(.data .data.*) 
          _edata = . ;         /* end of data constant  (from Xinu) */
       }                                   > ram
       .data1   : { *(.data1) }            > ram
       .rodata  : { *(.rodata .rodata.*) } > ram
       .rodata1 : { *(.rodata1) }          > ram
       .lit8    : { *(.lit8) }             > ram
       .lit4    : { *(.lit4) }             > ram
       .sdata   : { *(.sdata .sdata.*) }   > ram
       .sbss    : { *(.sbss .sbss.*) *(.scommon .scommon.*) }  > ram
       .bss :
       {
          *(.bss .bss.*) *(COMMON) 
          _end = . ;           /* end of image constant  (from Xinu) */
       }                                   > ram

       end_RAM = 0x; /* x_DATA_MEM_SZ */
       half_RAM = (end_RAM / 2);
       base_TP = ( _bdata + half_RAM );
       .TP base_TP : { *(.TP) }                  > ram
}
