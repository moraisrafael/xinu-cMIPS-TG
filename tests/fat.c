#include "cMIPS.h"

#ifndef cMIPS
  #include <stdio.h>
#endif

#ifdef cMIPS
extern void exit(int);
extern void print(int);
extern int readInt(int *);
extern void writeInt(int);
#endif

int myprint(int n, int f) {
  int i;
#ifdef cMIPS
  print(n);
  print(f);
#else
  printf("%08x:\n%08x:",n,f);
#endif
  return(n+f);
}

int fat(int n) {
   if (n==0) 
     return 1;
   else
     return (n * fat(n-1));
}

int fat2(int n) {
   int i; 
   for (i=(n-1); i>0; i--)
     n = n*i;
   return n;
}

int main() {
  int ret;

  //ret = myprint(3,fat2(3));
  //ret = myprint(3,fat(3));
  //ret = myprint(4,fat2(4));
  //ret = myprint(4,fat(4));
  //ret = myprint(8,fat(8));

  ret = myprint(7,fat2(7));  // 13b0
  ret = myprint(7,fat(7));   // 13b0

  ret = myprint(12,fat2(12));  //1c8cfc00
  ret = myprint(12,fat(12));

  return(ret);
}
