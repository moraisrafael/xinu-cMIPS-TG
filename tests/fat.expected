00000007
000013b0
00000007
000013b0
0000000c
1c8cfc00
0000000c
1c8cfc00
